File: AM/Lista2/Minesweeper/app/src/main/java/com/example/minesweeper/Grid.java
package com.example.minesweeper;

import java.util.ArrayList;
import java.util.List;
import java.util.Random;

public class Grid {
    private List<Cell> grid = new ArrayList<Cell>();
    private int size;

    public Grid(int size) {
        this.size = size;

        initializeGrid();
    }

    private void initializeGrid() {
        for (int i = 0; i < size * size; i++) {
            grid.add(new Cell(Cell.BLANK));
        }
    }

    public void generateGrid(int numberOfBombs) {
        int bombsPlaced = 0;
        while (bombsPlaced < numberOfBombs) {
            int x = new Random().nextInt(size);
            int y = new Random().nextInt(size);

            int index = toIndex(x, y);
            if (grid.get(index).getValue() == Cell.BLANK) {
                grid.set(index, new Cell(Cell.BOMB));
                bombsPlaced++;
            }
        }

        for (int x = 0; x < size; x++) {
            for (int y = 0; y < size; y++) {
                if (getCellAt(x, y).getValue() != Cell.BOMB) {
                    List<Cell> adjacentCells = getAdjacentCells(x, y);
                    int countBombs = 0;
                    for (Cell adjacentCell : adjacentCells) {
                        if (adjacentCell.getValue() == Cell.BOMB) {
                            countBombs++;
                        }
                    }

                    if (countBombs > 0) {
                        grid.set(toIndex(x, y), new Cell(countBombs));
                    }
                }
            }
        }
    }

    public List<Cell> getAdjacentCells(int x, int y) {
        List<Cell> adjacentCells = new ArrayList<Cell>();

        List<Cell> temporaryCellList = new ArrayList<Cell>();
        temporaryCellList.add(getCellAt(x - 1, y - 1));
        temporaryCellList.add(getCellAt(x, y - 1));
        temporaryCellList.add(getCellAt(x + 1, y - 1));
        temporaryCellList.add(getCellAt(x + 1, y));
        temporaryCellList.add(getCellAt(x + 1, y + 1));
        temporaryCellList.add(getCellAt(x, y + 1));
        temporaryCellList.add(getCellAt(x - 1, y + 1));
        temporaryCellList.add(getCellAt(x - 1, y));

        for (Cell cell : temporaryCellList) {
            if (cell != null) {
                adjacentCells.add(cell);
            }
        }

        return adjacentCells;
    }

    private Cell getCellAt(int x, int y) {
        if (x < 0 || x >= size || y < 0 || y >= size) {
            return null;
        }

        return grid.get(toIndex(x, y));
    }

    public int toIndex(int x, int y) {
        return x + (y * size);
    }

    public int[] toXY(int index) {
        int y = index / size;
        int x = index - (y * size);

        return new int[] { x, y };
    }

    public List<Cell> getCells() { return grid; }
}

File: AM/Lista2/MinesweeperAPK/app/src/main/java/com/example/minesweeper/Grid.kt
package com.example.minesweeper

import java.util.*

class Grid(private val size: Int) {
    private val grid: MutableList<Cell> = ArrayList()

    init {
        initializeGrid()
    }

    private fun initializeGrid() {
        for (i in 0 until size * size) {
            grid.add(Cell(Cell.BLANK))
        }
    }

    fun generateGrid(numberOfBombs: Int) {
        var bombsPlaced = 0
        while (bombsPlaced < numberOfBombs) {
            val x = Random().nextInt(size)
            val y = Random().nextInt(size)
            val index = toIndex(x, y)
            if (grid[index].value == Cell.BLANK) {
                grid[index] = Cell(Cell.BOMB)
                bombsPlaced++
            }
        }
        for (x in 0 until size) {
            for (y in 0 until size) {
                if (getCellAt(x, y)!!.value != Cell.BOMB) {
                    val adjacentCells = getAdjacentCells(x, y)
                    var countBombs = 0
                    for (adjacentCell in adjacentCells) {
                        if (adjacentCell.value == Cell.BOMB) {
                            countBombs++
                        }
                    }
                    if (countBombs > 0) {
                        grid[toIndex(x, y)] = Cell(countBombs)
                    }
                }
            }
        }
    }

    fun getAdjacentCells(x: Int, y: Int): List<Cell> {
        val adjacentCells: MutableList<Cell> = ArrayList()
        val temporaryCellList: MutableList<Cell?> = ArrayList()
        temporaryCellList.add(getCellAt(x - 1, y - 1))
        temporaryCellList.add(getCellAt(x, y - 1))
        temporaryCellList.add(getCellAt(x + 1, y - 1))
        temporaryCellList.add(getCellAt(x + 1, y))
        temporaryCellList.add(getCellAt(x + 1, y + 1))
        temporaryCellList.add(getCellAt(x, y + 1))
        temporaryCellList.add(getCellAt(x - 1, y + 1))
        temporaryCellList.add(getCellAt(x - 1, y))
        for (cell in temporaryCellList) {
            if (cell != null) {
                adjacentCells.add(cell)
            }
        }
        return adjacentCells
    }

    fun revealAllBombs() {
        for (cell in cells) {
            if (cell.value == Cell.BOMB) {
                cell.isRevealed = true
            }
        }
    }

    private fun getCellAt(x: Int, y: Int): Cell? {
        return if (x < 0 || x >= size || y < 0 || y >= size) {
            null
        } else grid[toIndex(x, y)]
    }

    fun toIndex(x: Int, y: Int): Int {
        return x + y * size
    }

    fun toXY(index: Int): IntArray {
        val y = index / size
        val x = index - y * size
        return intArrayOf(x, y)
    }

    val cells: List<Cell>
        get() = grid
}